include/envoy/grpc/async_client.h

```diff
@@ -0,0 +1,103 @@
+#pragma once
+
+#include <chrono>
+
+#include "envoy/common/optional.h"
+#include "envoy/common/pure.h"
+#include "envoy/http/header_map.h"
+
+namespace Envoy {
+namespace Grpc {
+
+/**
+ * Async client failure reasons.
+ */
+enum class AsyncClientFailureReason {
+  // The stream has been reset.
+  Reset,
+  // The stream has gRPC errors.
+  // TODO(htuch): Consider providing more granular gRPC failure information (e.g. status code,
+  // framing error, resource exhaustion, authentication failure).
+  GrpcError,
+};
+
+/**
+ * An in-flight gRPC stream.
+ */
+template <class RequestType> class AsyncClientStream {
+public:
+  virtual ~AsyncClientStream() {}
+
+  /**
+   * Send request message to the stream.
+   * @param request protobuf serializable message.
+   * @param end_stream supplies whether this is the last message.
+   */
+  virtual void sendMessage(const RequestType& request, bool end_stream) PURE;
+
+  /***
+   * Reset the stream.
+   */
+  virtual void reset() PURE;
+};
+
+/**
+ * Notifies caller of async gRPC stream status.
+ * Note the gRPC stream is full-duplex, even if the local to remote stream has been ended by
+ * Stream.sendMessage with end_stream=true, AsyncClientCallbacks can continue to receive events
+ * until the remote to local stream is closed, and vice versa.
+ * TODO(htuch): Add support for metadata.
+ */
+template <class ResponseType> class AsyncClientCallbacks {
+public:
+  virtual ~AsyncClientCallbacks() {}
+
+  /**
+   * Called when an async gRPC message is received.
+   * @param response the gRPC message.
+   */
+  virtual void onMessage(const ResponseType& message) PURE;
+
+  /**
+   * Called when an async gRPC stream is remotely closed with successful status.
+   * @param metadata the trailing metadata.
+   */
+  virtual void onRemoteClose() PURE;
```

Comment:
where is `metadata`? also there should be a callback to handle initial metadata too.